src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,518 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System;
+using static System.Math;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Quantum.Simulation.Common;
+using Microsoft.Quantum.Simulation.Core;
+using Microsoft.Quantum.Simulation.Simulators.Exceptions;
+
+namespace Microsoft.Quantum.Simulation.Simulators
+{
+
+    // ToffoliSimulator is implemented based on `QuantumProcessorDispatcher`,
+    // which is constructed using a specialization `QuantumProcessorBase`.  The
+    // specialization overrides methods to specify actions on intrinsic
+    // operations in the Q# code.
+    class ToffoliSimulatorProcessor : QuantumProcessorBase
+    {
+        // This property controls whether to throw an exception when a qubit is
+        // released that is not in the zero state.
+        public bool ThrowOnReleasingQubitsNotInZeroState { get; set; } = true;
+
+        // The simulation values are stored in the bits of a BitArray, which
+        // can use integers to store several bits at once.  We initialize the
+        // bit array to contain up to 64 bits and will dynamically grow it,
+        // whenever this capacity is exceeded.
+        public BitArray SimulationValues { get; private set; }
+
+        public ToffoliDumpFormat DumpFormat { get; set; } = ToffoliDumpFormat.Automatic;
+
+        // A property that must be assigned from the dispatcher to return the allocated
+        // qubit ids from the qubit manager
+        public Func<IEnumerable<long>>? GetAllocatedIds { get; set; }
+
+        // A property that must be assigned from the dispatcher to print a Q# message
+        public Action<String>? Message { get; set; }
+
+        private Dictionary<Qubit, Stack<bool>> borrowedQubitStates { get; } = new Dictionary<Qubit, Stack<bool>>();
```

Comment:
>Stack<bool> [](start = 34, length = 11)

Is there a way to implement it more efficiently? Maybe BitArray and StackPointer?

Reply:
Are you referring to `Dictionary` or `Stack` or both?